ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,285 @@
+     */
+    SnapshotCache() = delete;
+
+    /*!
+     * \brief Copy constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     *
+     * \param from The value to copy to this object.
+     */
+    SnapshotCache(const SnapshotCache& from) = delete;
+
+    /*!
+     * \brief Assignment operator.
+     *
+     * \note This operator is not implemented and should not be used.
+     *
+     * \param that The value to assign to this object.
+     *
+     * \return A reference to this object.
+     */
+    SnapshotCache& operator=(const SnapshotCache& that) = delete;
+
+    std::string d_object_name;
+
+    /*
+     * Data for tracking snapshots.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_snapshot_var;
+    SAMRAI::tbox::Pointer<SAMRAI::hier::VariableContext> d_ctx;
+    int d_snapshot_idx = IBTK::invalid_index;
+    SAMRAI::hier::IntVector<NDIM> d_gcw = 1;
+
+    /*!
+     * Vector of snapshots consisting of time points and hierarchies. Each hierarchy allocates data only with the patch
+     * data index stored by this class.
+     */
+    std::vector<SnapshotType> d_snapshots;
+
+    bool d_registered_for_restart = false;
+};
+
+/*!
+ * Update a snapshot. The data from this patch index is copied into an array of indices controlled by this
+ * object. This outputs an error if a snapshot with that time point within the specified tolerance can not be found.
+ *
+ * Calls to this function may use a different patch index than the one used in setSnapshot(), but the underlying
+ * data layout must be consistent. This data layout is specified by the variable given to the object in the
+ * constructor.
+ */
+void updateSnapshot(SnapshotCache& cache,
+                    int u_idx,
+                    double time,
+                    SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy,
+                    double tol = 1.0e-8);
+
+/*!
+ * Transfer the snapshot at the specified time value within a given tolerance to the patch index u_idx on the
+ * supplied patch hierarchy. The patch index u_idx must contain sufficient ghost cell width to perform the
+ * operations used by the refinement operator.
+ *
+ * Note this can cause a significant amount of communication if the supplied patch hierarchy has a different
+ * configuration of patches than the snapshot patch hierarchy.
+ *
+ * This function does not synchronize the data on the current hierarchy (i.e. no coarsening is performed). It
+ * transfers the snapshot using the refinement operator.
+ */
+void fillSnapshotOnHierarchy(const SnapshotCache& cache,
+                             int u_idx,
+                             double time,
+                             SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy,
+                             const std::string& snapshot_refine_type,
+                             double tol = 1.0e-8);
```

Comment:
Shouldn't this now be a utility function?

Reply:
I wasn't sure if we had a set style for placing stand-alone functions. I can pull these into a separate file. Although I thought we might want to let the compiler do it's job best and put all functions based on `SnapshotCache` into the same translation unit.

Reply:
we typically put these things in files like `ibmesh_utilities.h`.